File: test/harness.py
import os
import uuid
from aiohttp import ClientSession, web
from multidict import MultiDict

class Harness(object):
	scopes = {}
	dirname = os.path.dirname(os.path.realpath(__file__))

	def __init__(self, host = '0.0.0.0', port = 7777):
		self.host = host
		self.port = port
		self.app = web.Application()
		self.app.add_routes([
			web.get('/', self.index_html_handler),
			web.get('/{name}.js', self.test_js_handler),
			web.post('/callback/{id}', self.callback_handler),
			web.post('/test/{scope_id}', self.test_handler),
		])

	async def start(self):
		self.runner = web.AppRunner(self.app)
		await self.runner.setup()
		self.site = web.TCPSite(self.runner, self.host, self.port)
		await self.site.start()
		print('harness listening on http://%s:%s'%(self.host, self.port))

	async def stop(self):
		await self.runner.cleanup()

	async def index_html_handler(self, request):
		with open(os.path.join(self.dirname, 'index.html'), mode = 'rb') as file:
			return web.Response(body = file.read(), headers = {'Content-Type': 'text/html'})

	async def test_js_handler(self, request):
		with open(os.path.join(self.dirname, '{}.js'.format(request.match_info['name'])), mode = 'rb') as file:
			return web.Response(body = file.read(), headers = {'Content-Type': 'application/javascript; charset=utf-8'})

	async def callback_handler(self, request):
		scope_id, callback_id = request.match_info['id'].split('.')
		arguments = await request.json()
		scope = self.scopes[scope_id]
		scope[scope_id + '.' + callback_id] = {
			'headers': list(request.headers.items()),
			'arguments': arguments,
		}
		if not arguments:
			return web.json_response(None)
		for action in arguments:
			async with ClientSession() as session:
				arguments = []
				if 'arguments' in action:
					arguments = action['arguments'] or []
				async with session.post(action['url'], json = arguments) as response:
					pass
		return web.json_response(None)

	async def test_handler(self, request):
		scope_id = request.match_info['scope_id']
		arguments = await request.json()
		scope = {
			scope_id: {
				'arguments': arguments,
				'headers': list(request.headers.items()),
			}
		}
		self.scopes[scope_id] = scope
		for action in arguments:
			args = {'headers': [['Accept', 'application/json']]}
			if 'headers' in action:
				args['headers'] += action['headers']
			async with ClientSession(**args) as session:
				arguments = []
				if 'arguments' in action:
					arguments = action['arguments'] or []
				async with session.post(action['url'], json = arguments) as response:
					pass
		del self.scopes[scope_id]
		return web.json_response(scope)

if __name__ == '__main__':
	import asyncio
	import sys
	from contextlib import suppress
	host = '127.0.0.1'
	port = 7777
	if len(sys.argv) >= 2:
		host = sys.argv[1]
	if len(sys.argv) >= 3:
		port = int(sys.argv[2])
	harness = Harness(host = host, port = port)
	loop = asyncio.get_event_loop()
	loop.run_until_complete(harness.start())
	with suppress(KeyboardInterrupt):
		loop.run_forever()
	loop.run_until_complete(harness.stop())

File: test/harness.js
W3C = {};

W3C.test = function(name, payload, callback) {
	QUnit.test(name, function(assert) {
		var scope = String(Math.floor(Math.random() * 2147483647));
		var request = payload(id => { return window.location.protocol + '//' + window.location.host + window.location.pathname + 'callback/' + scope + '.' + id; });
		var done = assert.async();
		fetch('/test/' + scope, {
			method: 'POST',
			headers: {
				'Accept': 'application/json',
				'Content-Type': 'application/json'
			},
			body: JSON.stringify(request)
		}).then(response => {
			assert.ok(response.ok, 'HTTP response status ' + response.statusText);
			if (!response.ok) {
				done();
			}
			return response.json();
		}).then(response => {
			assert.ok(true, 'request body is ' + JSON.stringify(request));
			assert.ok(true, 'reponse body is ' + JSON.stringify(response));
			callback(id => { return response[id ? scope + '.' + id : scope]; }, assert);
			done();
		});
	});
};

W3C.get_traceparent = function(headers, assert) {
	var traceparents = [];
	headers.forEach(kv => {
		var name = kv[0];
		var value = kv[1];
		if (name.match(/^traceparent$/i)) {
			traceparents.push(value);
		}
	});
	assert.equal(traceparents.length, 1, 'expect 1 traceparent from the header, got ' + JSON.stringify(traceparents));
	var match = traceparents[0].match(/^([0-9a-f]{2})-([0-9a-f]{32})-([0-9a-f]{16})-([0-9a-f]{2})$/);
	match = match && {
		version: match[1],
		trace_id: match[2],
		span_id: match[3],
		trace_flags: match[4]
	};
	assert.ok(match, 'parse traceparent ' + JSON.stringify(match));
	return match;
};

W3C.get_tracestates = function(headers, assert) {
	var tracestates = [];
	headers.forEach(kv => {
		var name = kv[0];
		var value = kv[1];
		if (name.match(/^tracestate$/i)) {
			tracestates.push(value);
		}
	});
	assert.ok(true, 'got tracestate from header ' + JSON.stringify(tracestates));
	return tracestates;
};

W3C.test_single_request = function(name, url, headers, callback) {
	W3C.test(name, function(scope) {
		return [
			{url: url, headers: headers, arguments: [{url: scope('1'), arguments: []}]}
		];
	}, function(scope, assert) {
		var traceparent = W3C.get_traceparent(scope('1').headers, assert);
		var tracestates = W3C.get_tracestates(scope('1').headers, assert);
		callback(traceparent, tracestates, assert);
	});
};
